### ISSUE TITLE
Request to merge the customized new features into official version

### ISSUE THREAD

**Issue by Lina-LI-Na** [learnweb/moodle-mod_ratingallocate#207]
Dear Developers,
Thank you very much for your great support and response to my earlier [request](https://github.com/learnweb/moodle-mod_ratingallocate/issues/206).

We understand your concern about the extra customize and maintenance workload. If we work with Moodle partner Catalyst to develop the customized features, would it possible to merge these features into the official version of fair allocation?  

These new features could also benefit other users of the Moodle community and we are happy to pay the cost of the official version's maintenance of this part. The customized features include bulk upload choice in CSV file, add group restriction to each choice, upload attachment for each choice, and manually allocate student for each choice before the auto allocation.

Looking forward to your reply.

Thank you very much.

Best regards,
Lina

**Comment by NinaHerrmann** [comment]
Hey Lina, 
thanks for your information and the effort you put in! 
What do you mean by 
> we are happy to pay the cost of the official version's maintenance of this part
Does that mean that you pay catalyst for every moodle update or only for the versions which you are using? 
We try to get early bird batches for all our plugins so we maintain all versions which would be the goal. However, I can imagine you do not want to have those cost. 

Best regards, 
Nina

**Comment by Lina-LI-Na** [comment]
Hey Nina,
Thank you very much for your kind reply. 
We are happy to explore it with Catalyst financially before we can commit to it. 

May I double check with you what do you mean by
> We try to get early bird batches for all our plugins so we maintain all versions which would be the goal.

Does that mean all the historical versions and future new versions? 
If we can pay catalyst for the maintenance of all versions, would you be willing to accept our codes with the new features in the official plugin?

Many thanks for your great contributions and support again.

Best regards,
Lina

**Comment by NinaHerrmann** [comment]
Hey Lina, 

I only meant future versions. Early Bird badges mean we have a finished version of the plugin by the release date of the new version. (Which means having previous alpha/beta version testing) 
I would be happy if you can inform me if that is doable for you and catalyst. 

Cheers, 
Nina

**Comment by Lina-LI-Na** [comment]
Hey Nina,
Yes, we are happy to support all future versions.
Thank you very much. If your team is willing to accept our codes, we will start to work with Catalyst for the developement and keep you posted for any update.
Many thanks.

Cheers,
Lina

**Comment by danmarsden** [comment]
Hi Team,

just picking up the work proposed by @Lina-LI-Na 

I see many of the improvements they are after are already covered in issues like #37, #16 and a slight variation on #20 (allowing any file to be provided - eg a supplementary PDF) - so assuming you were happy with the structure of a pull request and it ticked all the boxes regarding tests/compliance etc. I'm guessing there's an interest in receiving a pull request with those improvements, 
but there's one improvement that I don't think is quite covered so I've just created #208 that covers this a bit.

We're still pricing the proposed work on our end but it would be good to hear if you are interested in the concept behind #208 and or/if there are any specific things you'd like to see in that work before you were interested in having it become part of your core release.

thanks!

**Comment by NinaHerrmann** [comment]
Hey @danmarsden,
generally, we are interested in the improvements. However, as you know reviewing takes a lot of time so it would be nice if we could have multiple pull requests for small improvements rather than one big one. 
Cheers, 
Nina

**Comment by danmarsden** [comment]
Thanks @NinaHerrmann - as you've probably seen we've started submitting some PR's related to this - feel free to close this particular issue off as the actual work will be covered in PR's or in specific issues.

I'm trying to make sure we submit the PR's in order and with small chunks to make it easier for you to review on your end - but please let me know if there's anything missing or anything we can do to make it easier for you to review.

After we get the pagination (#218) and group stuff (#219) into a state you're happy with we'll be looking at CSV import of choices (#16) and then the more complex manual allocation (#208) - we'll try to split the work from #208 into a few different PR's to make it easier to review.

**Comment by danmarsden** [comment]
Hi @NinaHerrmann - have you got anyone that will be able to review our PR's soon? - would be really good if we could address any feedback you might have before we run out of budget on our end - we really don't want to maintain a separate fork if we can avoid it!!

thanks!

**Comment by NinaHerrmann** [comment]
Hey @danmarsden thanks for splitting the PRs, I try to get as much as possible done today :+1:

**Comment by danmarsden** [comment]
Thanks for getting started on these @NinaHerrmann - great to see some of it merged in already! - We've got a csv import PR incoming soon and then will be looking into #208 which we expect to be a bit more complex.

thanks again!

**Comment by Laur0r** [comment]
Thank you for providing those features!
